tests/migrations/test_runner_individual.py

```diff
@@ -0,0 +1,235 @@
+        `contexts.key` Array(String), `contexts.value` Array(String), `_contexts_flattened` String,
+        `partition` UInt16, `offset` UInt64, `message_timestamp` DateTime, `retention_days` UInt16,
+        `deleted` UInt8) ENGINE = ReplacingMergeTree(deleted) PARTITION BY (retention_days, toMonday(finish_ts))
+        ORDER BY (project_id, toStartOfDay(finish_ts), transaction_name, cityHash64(span_id))
+        TTL finish_ts + toIntervalDay(retention_days);
+        """
+    )
+
+    assert get_sampling_key() == ""
+    generate_transactions()
+
+    runner = Runner()
+    runner.run_migration(MigrationKey(MigrationGroup.SYSTEM, "0001_migrations"))
+    runner._update_migration_status(
+        MigrationKey(MigrationGroup.TRANSACTIONS, "0001_transactions"), Status.COMPLETED
+    )
+    runner.run_migration(
+        MigrationKey(
+            MigrationGroup.TRANSACTIONS,
+            "0002_transactions_onpremise_fix_orderby_and_partitionby",
+        ),
+        force=True,
+    )
+
+    assert get_sampling_key() == "cityHash64(span_id)"
+
+    assert connection.execute("SELECT count(*) FROM transactions_local;") == [(5,)]
+
+
+def generate_transactions() -> None:
+    from datetime import datetime
+
+    table_writer = get_writable_storage(StorageKey.TRANSACTIONS).get_table_writer()
+
+    rows = []
+
+    for i in range(5):
+        raw_transaction = get_raw_transaction()
+        # Older versions of this table did not have measurements
+        del raw_transaction["data"]["measurements"]
+        del raw_transaction["data"]["breakdowns"]
+
+        processed = (
+            table_writer.get_stream_loader()
+            .get_processor()
+            .process_message(
+                (2, "insert", raw_transaction),
+                KafkaMessageMetadata(0, 0, datetime.utcnow()),
+            )
+        )
+        rows.extend(processed.rows)
+
+    BatchWriterEncoderWrapper(
+        table_writer.get_batch_writer(metrics=DummyMetricsBackend(strict=True)),
+        JSONRowEncoder(),
+    ).write(rows)
+
+
+def test_groupedmessages_compatibility() -> None:
+
+    # these three commands are repeated
+    cluster = get_cluster(StorageSetKey.EVENTS)
+    database = cluster.get_database()
+    connection = cluster.get_query_connection(ClickhouseClientSettings.MIGRATE)
+
+    # Create old style table witihout project ID
+    connection.execute(
+        """
+        CREATE TABLE groupedmessage_local (`offset` UInt64, `record_deleted` UInt8,
+        `id` UInt64, `status` Nullable(UInt8), `last_seen` Nullable(DateTime),
+        `first_seen` Nullable(DateTime), `active_at` Nullable(DateTime),
+        `first_release_id` Nullable(UInt64)) ENGINE = ReplacingMergeTree(offset)
+        ORDER BY id SAMPLE BY id SETTINGS index_granularity = 8192
+        """
+    )
+
+    migration_id = "0010_groupedmessages_onpremise_compatibility"
+
+    # Runner() instantiation, run_migration(), update status
```

Comment:
The three lines below are already saying the same thing very clearly. No need to add a comment to explain them.

Reply:
Same as above.